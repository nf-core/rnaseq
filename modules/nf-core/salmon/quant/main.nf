process SALMON_QUANT {
    tag "$meta.id"
    label "process_medium"

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/salmon:1.10.1--h7e5ed60_0' :
        'biocontainers/salmon:1.10.1--h7e5ed60_0' }"

    input:
    tuple val(meta), path(reads)
    path  index
    path  gtf
    path  transcript_fasta
    val   alignment_mode
    val   lib_type

    output:
    tuple val(meta), path("${prefix}") , emit: results
    tuple val(meta), path("*info.json"), emit: json_info, optional: true
    path  "versions.yml"               , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args   ?: ''
    prefix   = task.ext.prefix ?: "${meta.id}"

    def reference   = "--index $index"
    def reads1 = [], reads2 = []
    meta.single_end ? [reads].flatten().each{reads1 << it} : reads.eachWithIndex{ v, ix -> ( ix & 1 ? reads2 : reads1) << v }
    def input_reads = meta.single_end ? "-r ${reads1.join(" ")}" : "-1 ${reads1.join(" ")} -2 ${reads2.join(" ")}"
    if (alignment_mode) {
        reference   = "-t $transcript_fasta"
        input_reads = "-a $reads"
    }

    def strandedness_opts = [
        'A', 'U', 'SF', 'SR',
        'IS', 'IU' , 'ISF', 'ISR',
        'OS', 'OU' , 'OSF', 'OSR',
        'MS', 'MU' , 'MSF', 'MSR'
    ]
    def strandedness =  'A'
    if (lib_type) {
        if (strandedness_opts.contains(lib_type)) {
            strandedness = lib_type
        } else {
            log.info "[Salmon Quant] Invalid library type specified '--libType=${lib_type}', defaulting to auto-detection with '--libType=A'."
        }
    } else {
        strandedness = meta.single_end ? 'U' : 'IU'
        if (meta.strandedness == 'forward') {
            strandedness = meta.single_end ? 'SF' : 'ISF'
        } else if (meta.strandedness == 'reverse') {
            strandedness = meta.single_end ? 'SR' : 'ISR'
        }
    }
    """
    salmon quant \\
        --geneMap $gtf \\
        --threads $task.cpus \\
        --libType=$strandedness \\
        $reference \\
        $input_reads \\
        $args \\
        -o $prefix

    if [ -f $prefix/aux_info/meta_info.json ]; then
        cp $prefix/aux_info/meta_info.json "${prefix}_meta_info.json"
    fi

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        salmon: \$(echo \$(salmon --version) | sed -e "s/salmon //g")
    END_VERSIONS
    """
}
