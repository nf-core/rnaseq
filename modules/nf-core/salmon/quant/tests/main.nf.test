nextflow_process {

    name "Test Process SALMON_QUANT"
    script "../main.nf"
    process "SALMON_QUANT"
    config "./nextflow.config"
    tag "modules"
    tag "modules_nfcore"
    tag "salmon"
    tag "salmon/quant"
    tag "salmon/index"

    setup {
        run("SALMON_INDEX") {
            script "../../../salmon/index/main.nf"
            process {
                """
                input[0] = Channel.of([file(params.modules_testdata_base_path + "genomics/homo_sapiens/genome/genome.fasta", checkIfExists: true)])
                input[1] = Channel.of([file(params.modules_testdata_base_path + "genomics/sarscov2/genome/transcriptome.fasta", checkIfExists: true)])
                """
            }
        }
    }

    test("sarscov2 - single_end") {

        when {
            params {
                outdir = "$outputDir"
            }
            process {
                """
                input[0] = Channel.of([
                    [ id:'test', single_end: true ], // meta map
                    [ file(params.modules_testdata_base_path + "genomics/sarscov2/illumina/fastq/test_1.fastq.gz", checkIfExists: true) ]
                ])
                input[1] = SALMON_INDEX.out.index
                input[2] = Channel.of([file(params.modules_testdata_base_path + "genomics/sarscov2/genome/genome.gtf", checkIfExists: true)])
                input[3] = Channel.of([file(params.modules_testdata_base_path + "genomics/sarscov2/genome/transcriptome.fasta", checkIfExists: true)])
                input[4] = false
                input[5] = ''
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert path(process.out.json_info.get(0).get(1)).exists() },
                { assert path(process.out.results.get(0).get(1)).exists() },
                { assert snapshot(process.out.versions).match("versions_single_end") }
            )
        }

    }

    test("sarscov2 - single_end lib type A") {

        when {
            params {
                outdir = "$outputDir"
            }
            process {
                """
                input[0] = Channel.of([
                    [ id:'test', single_end: true ], // meta map
                    [ file(params.modules_testdata_base_path + "genomics/sarscov2/illumina/fastq/test_1.fastq.gz", checkIfExists: true) ]
                ])
                input[1] = SALMON_INDEX.out.index
                input[2] = Channel.of([file(params.modules_testdata_base_path + "genomics/sarscov2/genome/genome.gtf", checkIfExists: true)])
                input[3] = Channel.of([file(params.modules_testdata_base_path + "genomics/sarscov2/genome/transcriptome.fasta", checkIfExists: true)])
                input[4] = false
                input[5] = 'A'
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert path(process.out.json_info.get(0).get(1)).exists() },
                { assert path(process.out.results.get(0).get(1)).exists() },
                { assert snapshot(process.out.versions).match("versions_single_end_lib_type_a") }
            )
        }

    }

    test("sarscov2 - pair_end") {

        when {
            params {
                outdir = "$outputDir"
            }
            process {
                """
                input[0] = Channel.of([
                    [ id:'test', single_end: true ], // meta map
                    [ 
                        file(params.modules_testdata_base_path + "genomics/sarscov2/illumina/fastq/test_1.fastq.gz", checkIfExists: true),
                        file(params.modules_testdata_base_path + "genomics/sarscov2/illumina/fastq/test_2.fastq.gz", checkIfExists: true)
                    ]
                ])
                input[1] = SALMON_INDEX.out.index
                input[2] = Channel.of([file(params.modules_testdata_base_path + "genomics/sarscov2/genome/genome.gtf", checkIfExists: true)])
                input[3] = Channel.of([file(params.modules_testdata_base_path + "genomics/sarscov2/genome/transcriptome.fasta", checkIfExists: true)])
                input[4] = false
                input[5] = ''
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert path(process.out.json_info.get(0).get(1)).exists() },
                { assert path(process.out.results.get(0).get(1)).exists() },
                { assert snapshot(process.out.versions).match("versions_pair_end") }
            )
        }

    }

    test("sarscov2 - pair_end multiple") {

        when {
            params {
                outdir = "$outputDir"
            }
            process {
                """
                input[0] = Channel.of([
                    [ id:'test', single_end:false ], // meta map
                    [
                        file(params.test_data['sarscov2']['illumina']['test_1_fastq_gz'], checkIfExists: true),
                        file(params.test_data['sarscov2']['illumina']['test_2_fastq_gz'], checkIfExists: true),
                        file(params.test_data['sarscov2']['illumina']['test2_1_fastq_gz'], checkIfExists: true),
                        file(params.test_data['sarscov2']['illumina']['test2_2_fastq_gz'], checkIfExists: true)
                    ]
                ])
                input[1] = SALMON_INDEX.out.index
                input[2] = Channel.of([file(params.modules_testdata_base_path + "genomics/sarscov2/genome/genome.gtf", checkIfExists: true)])
                input[3] = Channel.of([file(params.modules_testdata_base_path + "genomics/sarscov2/genome/transcriptome.fasta", checkIfExists: true)])
                input[4] = false
                input[5] = ''
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert path(process.out.json_info.get(0).get(1)).exists() },
                { assert path(process.out.results.get(0).get(1)).exists() },
                { assert snapshot(process.out.versions).match("versions_pair_end_multiple") }
            )
        }

    }
}
