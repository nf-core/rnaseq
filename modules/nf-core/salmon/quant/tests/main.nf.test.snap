{
    "versions_single_end": {
        "content": [
            [
                "versions.yml:md5,80eb3d2ad36960c7e9263f81ede9d263"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-02-01T11:51:01.865730832"
    },
    "versions_pair_end": {
        "content": [
            [
                "versions.yml:md5,80eb3d2ad36960c7e9263f81ede9d263"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-02-06T17:10:56.121713"
    },
    "versions_single_end_lib_type_a": {
        "content": [
            [
                "versions.yml:md5,80eb3d2ad36960c7e9263f81ede9d263"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-02-06T17:09:14.898351"
    },
    "versions_pair_end_multiple": {
        "content": [
            [
                "versions.yml:md5,80eb3d2ad36960c7e9263f81ede9d263"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-02-06T17:11:07.846643"
    }
}